### DISCUSSION TITLE
NetworkFailure/ApiFailure and Moshi exceptions?

### DISCUSSION THREAD

**Discussion by mlewandowskipt** [slackhq/EitherNet#84] [Q&A] [answered]
### Description

Hi, in my project I have following Retrofit instance:
```kotlin
val retrofit = Retrofit.Builder()
    .baseUrl(...)
    .addConverterFactory(ApiResultConverterFactory)
    .addConverterFactory(MoshiConverterFactory.create(moshi))
    .addCallAdapterFactory(ApiResultCallAdapterFactory)
    .client(okHttpClient)
    .build()
```

When there's some JSON malformation in the response, Moshi's converter throws `JsonDataException` or `JsonEncodingException` which extend `IOException`. For this reason `ApiResultCallAdapter` transforms the exception into `NetworkFailure`, which is wrong, since it's not a network problem but more an API problem.

What is the recommended way to handle this? I'd like to transform those Moshi's exceptions into `ApiFailure` and pass the original exception as its `error` to report it to crash reporting service, but I think there's no straightforward way to do this. The readme only mentions how to throw `ApiException` in case of some specific response, but not in the case when some other converter threw an exception.

It's more a question if there's a suggested way how to handle this, or maybe there would be a possibility to introduce some parameter to `ApiResultCallAdapterFactory` that would change how specific exception types are mapped.

**Comment by ZacSweers** [comment, marked as answer]
You would need to write your own intercepting JsonAdapter to capture those and rethrow them with your own exception. This is what we do in Slack's network stack and the intended pattern for this scenario in EitherNet.
